### DISCUSSION TITLE
[Lock & Semaphore] Why is the lock state based on the instance of the Lock object?

### DISCUSSION THREAD

**Discussion by nusje2000** [symfony/symfony#45001] [Q&A] [answered]
I was using the lock component and wondered why the lock instance is bound to the acquired lock. I expected the lock component to create a lock based on the resource provided when creating the lock using the `LockFactory`.

Example:
```php
/** @var LockFactory $factory */

$instance1 = $factory->createLock('resource_x');
$instance1->acquire(); // will return true, because no lock has been created yet
$instance1->isAcquired(); // will return true, because the line above has acquired a lock

$instance2 = $factory->createLock('resource_x');
$instance2->isAcquired(); // will return false because the Lock instance is not the same
```

I was creating a messenger transport that utilizes the lock to lock certain resources from being processed by other consumers. Both the get, ack and reject created a new instance of the lock and therefore the lock was never released in the ack and reject methods, because they didn't use the same lock instance that the get method used when acquiring the lock:

```php
class SomeLockingTrasport implements TransportInterface
{
    private LockFactory $lockFactory;

    public function get(): iterable
    {
        // logic to retrieve a resource

        if ($this->lockFactory->createLock($resource->identifier())->acquire()) {
            return [$resource];
        }
        
        return [];
    }

    public function ack(Envelope $envelope): void
    {
        // logic for acknowledging the processed envelope

        // never releases the lock because it does not use the same instance as the get() method
        $this->lockFactory->createLock($resource->identifier())->release(); 
    }

    public function reject(Envelope $envelope): void { /* same issue as described in the ack() method */ }

    public function send(Envelope $envelope): Envelope {/* logic for sending the message */}
}
```

Although storing the lock in a property and later using that same instance for releasing the lock does work, I then still encounter the same issues when writing integration tests for this code where I don't have access to the original lock instance.

I understand how the locking mechanism works, and the "issues" that I did encounter are well documented. But I don't really understand *why* the lock component behaves this way (same goes for the semaphore component).

If there is anyone that can explain this to me, I would really appreciate it.

**Comment by stof** [comment, marked as answer]
Well, creating a state based on the resource would totally defeat the purpose of the component, as it would never lock anything. The whole purpose of locking is to prevent multiple attempts to lock the same resource in parallel.

So you need to use the same Lock instance (or at least the same Key instance) to acquire and release the logic.

Side note: your transport is flawed. The default behavior of the Lock class is to release the lock automatically on `__destruct` (which is what needs to be disabled when wanting to transfer the `Key` around for advanced used cases), and so you code doing `if ($this->lockFactory->createLock($resource->identifier())->acquire()) {` is releasing its lock immediately after acquiring it (as the lock object is not kept around).

**Reply by nusje2000** [reply]
@stof thanks for your explanation.

In reply to the sidenote: good point, I forgot the auto release=false in my example.

**Reply by nusje2000** [reply]
What would be a correct way to preserve the key (or lock) in this example. Does it make sense to attach the key to the message, or maybe add a stamp to the envelope?

**Reply by stof** [reply]
I don't know whether using a stamp or putting it in the message makes more sense. I don't have enough context for that.
